File: pec-observation-app/ios-app/Sources/PECObservation/Application/UseCases/SyncUseCase.swift
import Foundation

public struct PresignedURLResponse: Decodable {
    public let uploadId: String
    public let uploadUrl: String
    public let expiresAt: String
}

public struct ConfirmUploadResponse: Decodable {
    public let uploadId: String
    public let status: String
}

public final class SyncUseCase {
    private let syncQueue: SyncQueuePort
    private let network: NetworkClientPort
    private let token: () -> String?
    private let maxRetries = 5

    public init(syncQueue: SyncQueuePort, network: NetworkClientPort, token: @escaping () -> String?) {
        self.syncQueue = syncQueue
        self.network = network
        self.token = token
    }

    public func processPendingCommands() async {
        let commands: [SyncCommand]
        do {
            commands = try syncQueue.pendingCommands()
        } catch {
            return
        }

        for command in commands {
            guard command.retryCount < maxRetries else {
                try? syncQueue.markFailed(command.commandId, reason: "Max retries exceeded")
                continue
            }
            do {
                try? syncQueue.markInProgress(command.commandId)
                try await processCommand(command)
                try? syncQueue.markCompleted(command.commandId)
            } catch {
                let delay = pow(2.0, Double(command.retryCount)) // 1, 2, 4, 8, 16 seconds
                try? syncQueue.markFailed(command.commandId,
                                          reason: error.localizedDescription)
                try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))
            }
        }
    }

    private func processCommand(_ command: SyncCommand) async throws {
        switch command.type {
        case .uploadAudio:
            try await processAudioUpload(command)
        case .createObservation, .updateObservation, .createSchool, .createTeacher:
            // These are handled via direct API calls in the view models;
            // sync queue is used for retry if offline
            break
        }
    }

    private func processAudioUpload(_ command: SyncCommand) async throws {
        guard let observationId = command.payload["observationId"],
              let audioLocalPath = command.payload["audioLocalPath"],
              let audioURL = URL(string: audioLocalPath) else {
            throw DomainError.validationFailed("Invalid upload audio command payload")
        }

        let audioData = try Data(contentsOf: audioURL)
        let fileSize = command.payload["fileSizeBytes"].flatMap { Int($0) } ?? audioData.count
        let checksum = command.payload["checksum"] ?? ""
        let codec = command.payload["codec"] ?? "aac"
        let duration = command.payload["durationSeconds"].flatMap { Int($0) } ?? 0

        let presignedBody = [
            "observation_id": observationId,
            "file_name": "observation-\(observationId).m4a",
            "file_size_bytes": fileSize,
            "checksum": checksum,
            "codec": codec,
            "idempotency_key": command.idempotencyKey.uuidString,
        ] as [String: Any]

        let presignedResponse: PresignedURLResponse = try await network.post(
            path: "/api/v1/audio/presigned-url",
            body: AnyEncodable(presignedBody),
            token: token()
        )

        guard let uploadURL = URL(string: presignedResponse.uploadUrl) else {
            throw NetworkError.decodingFailed("Invalid presigned URL")
        }
        try await network.put(url: uploadURL, data: audioData, contentType: "audio/mp4")

        let _: ConfirmUploadResponse = try await network.post(
            path: "/api/v1/audio/confirm/\(presignedResponse.uploadId)",
            body: ["duration_seconds": duration],
            token: token()
        )
    }
}

// Helper to encode arbitrary dictionaries
struct AnyEncodable: Encodable {
    private let _encode: (Encoder) throws -> Void

    init(_ value: [String: Any]) {
        self._encode = { encoder in
            var container = encoder.singleValueContainer()
            let jsonData = try JSONSerialization.data(withJSONObject: value)
            let decoded = try JSONDecoder().decode(AnyCodable.self, from: jsonData)
            try container.encode(decoded)
        }
    }

    func encode(to encoder: Encoder) throws {
        try _encode(encoder)
    }
}

private struct AnyCodable: Codable {
    private var value: Any

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let v = try? container.decode([String: AnyCodable].self) {
            value = v
        } else if let v = try? container.decode([AnyCodable].self) {
            value = v
        } else if let v = try? container.decode(String.self) {
            value = v
        } else if let v = try? container.decode(Int.self) {
            value = v
        } else if let v = try? container.decode(Double.self) {
            value = v
        } else if let v = try? container.decode(Bool.self) {
            value = v
        } else {
            value = ""
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch value {
        case let v as [String: AnyCodable]:
            try container.encode(v)
        case let v as [AnyCodable]:
            try container.encode(v)
        case let v as String:
            try container.encode(v)
        case let v as Int:
            try container.encode(v)
        case let v as Double:
            try container.encode(v)
        case let v as Bool:
            try container.encode(v)
        default:
            try container.encodeNil()
        }
    }
}

File: pec-observation-app/android-app/app/src/main/kotlin/br/gov/educacao/sp/pecobservation/application/usecases/SyncUseCase.kt
package br.gov.educacao.sp.pecobservation.application.usecases

import android.util.Log
import br.gov.educacao.sp.pecobservation.domain.entities.SyncCommandType
import br.gov.educacao.sp.pecobservation.ports.NetworkClientPort
import br.gov.educacao.sp.pecobservation.ports.SyncQueuePort
import kotlinx.coroutines.delay
import java.util.UUID
import javax.inject.Inject

private const val TAG = "SyncUseCase"
private const val MAX_RETRIES = 5

class SyncUseCase @Inject constructor(
    private val syncQueue: SyncQueuePort,
    private val network: NetworkClientPort,
) {
    suspend fun execute(): SyncResult {
        val pending = syncQueue.pendingCommands().sortedBy { it.createdAt }
        var succeeded = 0
        var failed = 0

        for (command in pending) {
            syncQueue.markInProgress(command.commandId)
            try {
                when (command.type) {
                    SyncCommandType.CREATE_OBSERVATION -> {
                        val obsId = command.payload["observationId"]!!
                        network.post("/api/v1/observations", mapOf("id" to obsId))
                    }
                    SyncCommandType.UPDATE_OBSERVATION -> {
                        val obsId = command.payload["observationId"]!!
                        network.patch("/api/v1/observations/$obsId", command.payload)
                    }
                    SyncCommandType.CREATE_SCHOOL -> {
                        network.post("/api/v1/schools", command.payload)
                    }
                    SyncCommandType.CREATE_TEACHER -> {
                        network.post("/api/v1/teachers", command.payload)
                    }
                    SyncCommandType.UPLOAD_AUDIO -> processAudioUpload(command.payload)
                }
                syncQueue.markCompleted(command.commandId)
                succeeded++
            } catch (e: Exception) {
                Log.w(TAG, "Command ${command.commandId} failed: ${e.message}")
                val retries = command.retryCount
                if (retries < MAX_RETRIES) {
                    delay(exponentialBackoffMs(retries))
                    syncQueue.markFailed(command.commandId, e.message ?: "unknown")
                } else {
                    syncQueue.markFailed(command.commandId, "max retries exceeded: ${e.message}")
                    failed++
                }
            }
        }
        return SyncResult(succeeded, failed)
    }

    private suspend fun processAudioUpload(payload: Map<String, String>) {
        val observationId = payload["observationId"]!!
        val localFilePath = payload["localFilePath"]!!
        val checksum = payload["checksum"]!!

        val presignedResponse = network.post(
            "/api/v1/audio/presigned-url",
            mapOf("observation_id" to observationId, "checksum" to checksum),
        )
        val uploadUrl = presignedResponse["upload_url"] as String
        val uploadId  = presignedResponse["upload_id"] as String

        network.putBinary(uploadUrl, localFilePath)
        network.post("/api/v1/audio/confirm/$uploadId", mapOf("checksum" to checksum))
    }

    private fun exponentialBackoffMs(retryCount: Int): Long =
        (1_000L * (1 shl retryCount)).coerceAtMost(32_000L)
}

data class SyncResult(val succeeded: Int, val failed: Int)
